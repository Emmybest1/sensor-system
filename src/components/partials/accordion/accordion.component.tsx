import React from 'react';
import styled from 'styled-components';

import { useAccordion } from './accordion.hook';

/** ******
 * <Accordion.Header/>
 */
const HeaderWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

const TitleWrapper = styled.h3`
  font-size: 0.9rem;
  font-weight: 300;
`;

const DropdownIconWrapper = styled.div`
  background-color: #676a71;
  border-radius: 50px;
  height: 30px;
  width: 30px;
  display: flex;
  justify-content: center;
  align-items: center;
  align-content: center;
  transition: all 0.5s ease-out;

  &.accordion--closed {
    transform: rotate(180deg);
  }
`;

type HeaderPropTypes = {
  isOpened?: boolean;
  title: string;
  toggleIsOpened?: React.MouseEventHandler;
};

const Header: React.FC<HeaderPropTypes> = ({ title, isOpened, toggleIsOpened }) => (
  <HeaderWrapper>
    <TitleWrapper>
      <span className="panel-hash">#</span>&nbsp;{title}
    </TitleWrapper>
    <DropdownIconWrapper role="button" className={`accordion ${isOpened ? 'accordion--opened' : 'accordion--closed'}`} onClick={toggleIsOpened}>
      <i className="fa fa-angle-up" aria-hidden="true" />
    </DropdownIconWrapper>
  </HeaderWrapper>
);

const BodyWrapper = styled.div`
  padding-left: 15px;
  font-size: 0.8rem;

  & > * {
    color: #949494;

    &:hover {
      color: #eee;
    }
  }
`;

/** ******
 * <Accordion.Body/>
 */
type BodyPropTypes = {
  isOpened?: boolean;
  children: React.ReactNode;
};

const Body: React.FC<BodyPropTypes> = ({ children, isOpened }) =>
  isOpened ? (
    <BodyWrapper>
      <span className="panel-hash">#</span>&nbsp;{children}
    </BodyWrapper>
  ) : null;

/** ******
 * <Accordion/>
 */
const AccordionWrapper = styled.div``;

type AccordionPropTypes = {
  Header: React.FC<HeaderPropTypes>;
  Body: React.FC<BodyPropTypes>;
};

export const Accordion: React.FC & AccordionPropTypes = ({ children }) => {
  const { isOpened, toggleIsOpened } = useAccordion();

  const newNode = React.Children.map(children, (child) => {
    const item = child as React.ReactElement<React.PropsWithChildren<HeaderPropTypes & BodyPropTypes>>;

    if (item.type === Header) return React.cloneElement(item, { toggleIsOpened, isOpened });

    if (item.type === Body) return React.cloneElement(item, { isOpened });

    return child;
  });

  return <AccordionWrapper>{newNode}</AccordionWrapper>;
};

Accordion.Header = Header;
Accordion.Body = Body;
