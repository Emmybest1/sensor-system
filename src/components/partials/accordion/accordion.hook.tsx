import { useState } from 'react';

export const useAccordion = () => {
  const [isOpened, setIsOpened] = useState<boolean>();

  const toggleIsOpened = () => {
    setIsOpened((prevIsOpened) => !prevIsOpened);
  };

  return { isOpened, toggleIsOpened };
};
