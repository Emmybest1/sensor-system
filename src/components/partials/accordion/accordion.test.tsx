import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { Accordion } from './accordion.component';

describe('<Accordion/>', () => {
  test('Should render Component', () => {
    const { container } = render(
      <Router>
        <Accordion>
          <Accordion.Header title="Header" />
          <Accordion.Body>
            <ul>
              <li>Item 1</li>
              <li>Item 2</li>
              <li>Item 3</li>
            </ul>
          </Accordion.Body>
        </Accordion>
      </Router>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
