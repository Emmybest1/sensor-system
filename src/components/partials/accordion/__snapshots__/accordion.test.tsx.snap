// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion/> Should render Component 1`] = `
<div
  class="sc-eCstlR"
>
  <div
    class="sc-bdfBQB fNVPBu"
  >
    <h3
      class="sc-gsTEea gpJJjB"
    >
      <span
        class="panel-hash"
      >
        #
      </span>
       
      Header
    </h3>
    <div
      class="sc-dlfnuX bYesIP accordion accordion--opened"
      role="button"
    >
      <i
        aria-hidden="true"
        class="fa fa-angle-up"
      />
    </div>
  </div>
  <div
    class="sc-hKgJUU blJQxl"
  >
    <span
      class="panel-hash"
    >
      #
    </span>
     
    <ul>
      <li>
        Item 1
      </li>
      <li>
        Item 2
      </li>
      <li>
        Item 3
      </li>
    </ul>
  </div>
</div>
`;
